### PR TITLE
Add crash detection tool

### PR DESCRIPTION
**Pull request by ZvonimirMedak**
## Summary

This PR adds a crash detection tool, it will catch exceptions thrown on the system level, and save them into 

## Changes

Added new files for the CrashDetection tool

### Type

- [X] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [ ] **Refactor**: This pull request refactors existing code.
- [ ] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [ ] This pull request introduces a **breaking change**.

### Description

I have added the crash detection tool for which I got the inspiration from DebugSwift. We use it on our project, and i has proven really useful in combination with Pulse which we use for logging.
I have also updated the Readme, and DocC

## Checklist

- [X] I have performed a self-review of my own code.
- [X] I have tested my changes, including edge cases.
- [X] I have added necessary tests for the changes introduced (if applicable).
- [X] I have updated the documentation to reflect my changes (if applicable).

## Additional notes

Up for a debate if we should introduce it in the default subspec.
I have crashed the App by trying to access the third element of an empty array when creating the Preference items.

https://github.com/user-attachments/assets/b83e0659-90ff-4520-a08c-56418c18b19f

https://github.com/user-attachments/assets/2489700d-ccd6-4957-9e1f-f9f624af2e61

### CODE REVIEW

File: Sentinel/Classes/CrashDetection/Internal/CrashManager.swift [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+//  Sentinel
+//
+//  Created by Zvonimir Medak on 10.03.2025..
+//
+
+// Inspiration link: https://github.com/DebugSwift/DebugSwift
+
+import Foundation
+
+enum CrashManager {
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/CrashManager.swift:12]: Based on the functionality, I would say this is more like a crash file/data manager, and if so, I would rather change the name to `CrashDataManager` or `CrashFileManager` because I though this does more than file handling just by the name.

File: Sentinel/Classes/CrashDetection/Internal/Handlers/CrashHandler.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    init() {
+        let signalHandler = SignalPrehandlerManager()
+        uncaughtExceptionHandler = CrashUncaughtExceptionHandler(preHandlers: signalHandler.preHandlers)
+        signalExceptionHandler = CrashSignalExceptionHandler(preHandlers: signalHandler.preHandlers)
+
+        CrashUncaughtExceptionHandler.exceptionReceiveClosure = { [weak self] signal, exception, info in
+            self?.exceptionReceiveClosure?(signal, exception, info)
+            CrashManager.save(
+                crash: .init(type: .nsexception, details: .builder(name: info))
+            )
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Handlers/CrashHandler.swift:36]: Even though I'm fine with this, I would rather "register" or "add" result handler and then call it here instead of directly using this manager.

Reason; if we want to use another way of saving or previewing data for the user in the future (e.g., saving it to the database).

Same goes for the signal exception handler couple of lines below.

File: Sentinel/Classes/CrashDetection/Internal/Handlers/CrashHandler.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        CrashSignalExceptionHandler.exceptionReceiveClosure = { [weak self] signal, exception, info in
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Handlers/CrashHandler.swift:39]: Also, regarding `CrashUncaughtExceptionHandler` and `CrashSignalExceptionHandler`, it would be great if a developer can check/uncheck which exception handlers should be enabled and that will be saved based on that choice.

File: Sentinel/Classes/CrashDetection/Internal/Handlers/Preperable.swift [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+//
+//  Created by Zvonimir Medak on 18.04.2024..
+//  Copyright © 2024 Signify. All rights reserved.
+//
+
+// Inspiration link: https://github.com/DebugSwift/DebugSwift
+
+import Foundation
+
+protocol Preperable {
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Handlers/Preperable.swift:13]: Typo; `Prepareable`, even though, I would make the name less generic (in the scope of this tool only), e..g. `CrashExceptionPrepareable`.

File: Sentinel/Classes/CrashDetection/Internal/Trace.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+//
+//  Created by Zvonimir Medak on 10.03.2025..
+//
+
+// Inspiration link: https://github.com/DebugSwift/DebugSwift
+
+import Foundation
+import MachO
+
+enum Trace {
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Trace.swift:13]: If this is a stack trace, I would rather to with `StackTrace`.

File: Sentinel/Classes/CrashDetection/CrashToolDetailsView.swift [hunk 6]
```diff
@@ -0,0 +1,85 @@
+//
+//  CrashToolDetailsView.swift
+//  Sentinel
+//
+//  Created by Zvonimir Medak on 10.03.2025..
+//
+
+import SwiftUI
+
+struct CrashToolDetailsView: View {
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/CrashToolDetailsView.swift:10]: I would rather change names of this to `CrashDetectionToolDetailsView`, the same goes for the `CrashToolView`. I'm trying to avoid nitpicking, but this is more about scoping the functionality and use the naming in that context.

File: Sentinel/Classes/CrashDetection/Internal/Trace.swift [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    static func slideAddressForImageIndex(_ index: UInt32) -> Int64 {
+        return Int64(_dyld_get_image_vmaddr_slide(index))
+    }
+}
+
+private extension Trace {
+
+    @available(iOS, deprecated: 16, message: "We can use Regex")
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Trace.swift:64]: Make the message more actionable; e.g., "Use Regex instead."

File: Sentinel/Classes/CrashDetection/Internal/Handlers/CrashUncaughtExceptionHandler.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    func prepare() {
+        preUncaughtExceptionHandler = NSGetUncaughtExceptionHandler()
+        NSSetUncaughtExceptionHandler(UncaughtExceptionHandler)
+    }
+}
+
+func UncaughtExceptionHandler(exception: NSException) {
+    let reason = exception.reason ?? ""
```
Comment (nikolamajcen) [inline Sentinel/Classes/CrashDetection/Internal/Handlers/CrashUncaughtExceptionHandler.swift:31]: Put "Unknown" or "Undefined" instead.

**Review by nikolamajcen — CHANGES_REQUESTED** [review]
Nice work! I really like the idea and the overall implementation. I've added some comments, please check them out.

Overall, great 👍

**Review by nikolamajcen — APPROVED** [review]
All good, nice work! 👍
